restructure verifier to simplify pubsub and logic tracing

deitch opened the pull request:
## The change

This changes no functionality or business logic, but does a restructure of verifier, i.e. `pkg/pillar/cmd/verifier/`. The goal is to redo all ~30 `pkg/pillar/cmd/*` services in the same way.

* clean separation of business logic, library surface for calling from zedbox, pubsub handlers, and CLI
* creation of a CLI option, so you can run verifier as a small standalone binary, not just as part of zedbox
* when running standalone, it can be run to connect to an existing pubsub, by providing the directory, or just the business logic, e.g. `verifier --infile somefile --digest sha256:ababab1111 --mediaType ...` etc. It runs once and exits
* moves a lot of functions that were simple bare, e.g. `func handleInit()` but expected certain package-wide variables to be set, into members of a struct, e.g. `func (ctx *verifierContext) Init()`.
* As a result of the above, eliminates all package-wide variables, favouring structs

As a result of the above, it is possible to build and run a verifier binary that has almost no dependencies other than the things it actually includes. The compiled binary is 17MB, and probably could be made smaller with more effort. I have successfully compiled it for macos, linux and windows, and run it on all of the above.

Obviously, included a README describing the structure.

## Why?

Why are we doing this change? There is a number of benefits.

First, the business logic is cleanly separated from any other logic such as pubsub. This makes it much easier to understand as well as test the basic functions, and test any pubsub handlers by mocking the business logic functions. When the business logic functions really do _just_ business logic, that is much easier to do.

Second, it makes it possible to run it standalone, exercising just the business logic, or with pubsub. This leads to a much faster development cycle and manual testing, when needed. I now can run verifier _on a mac_ without any EVE running, feed it a message, and watch it succeed or fail.

Third, it forces us to look at the dependency tree. When verifier is part of zedbox, the final binary includes all dependencies. It is impossible to tease out if a package really needs to be included by the compiler and linker or not, as it could come from multiple places. With a binary that is _just_ verifier, we can examine the final binary and see what is included and reason why. This quickly finds large dependencies that can be factored out. 

For example, in this method, I found that verifier was including 2 functions in `utils/wait`. A third function depended on `pkg/pillar/containerd` which depended on `github.com/containerd/containerd` and therefore all of its dependencies. verifier doesn't need any of that. It allowed us to move the `WaitForContainerd` into `pkg/pillar/containerd` and not get it by accident. The size of verifier shrunk by 9MB (!!). As we do this for more and more services, we will be able to see which packages are unusually big, if they are needed, and tease out how to reduce zedbox binary size and memory usage.

Fourth, this prepares the groundwork for rebuilding some of these services in more memory-efficient languages or restructures in go. We do want to isolate some of these, e.g. those with access to the Internet such as downloader or zedagent, using Linux-level process isolation, network namespaces, capabilities, etc. This prepares the groundwork.

Fifth, this prepares the groundwork for a possible restructure of pubsub into something more off the shelf.

## Will we run separate binaries?

The goal for now remains to run verifier in EVE as a goroutine in zedbox, or at most as a single zedbox binary which is invoked multiple times, each time calling a different function. This opens up options in the future, as well as during development cycles.

## What is missing?

The only piece missing here is documentation. I added for verifier in `pkg/pillar/cmd/verifier/README.md`, but this deserves a more detailed, "here is how to write services". I will add that as a separate commit in this PR.

cc @uncleDecart @eriknordmark

deitch commented:
To repeat: this changes no logic, but restructures the calling of functions that do that logic. Thus, it is really important that testing is clean.

deitch commented:
Documentation added. Ready for some reviews.

uncleDecart commented on pkg/pillar/cmd/README.md:
```diff
@@ -0,0 +1,273 @@
+* [`lib/`](./lib/) contains core logic, has no dependency on being called command-line or pubsub, unless pubsub is part of the core business logic. It is important to keep these functions as simple and straightforward as possible. Unit tests of business logic should be here as well.
+* [`pubsub/`](./pubsub/) contains functions to handle pubsub messages, including all handlers. There should be no business logic here, only pubsub message handling. For business logic, handlers should call business logic functions in `lib/`.
+* [`cmd/`](./cmd/) contains CLI commands and sub-commands, all in package `main`. Can be called as `go run ./cmd`. Call either functions in `pubsub/` or `lib/` depending on the command.
+* [`run.go`](./run.go) is the entrypoint for calling as part of zedbox, exposes the implementation of [`types.AgentRunner`](https://pkg.go.dev/github.com/lf-edge/eve/pkg/pillar/types#AgentRunner).
```
Should we add note here that it's desirable architecture, however if it doesn't fit changes are possible? I'm referring here to metadata server (in current iteration of refactoring :art: )

deitch replied:
Yeah. I'm away from laptop, but I'll add that later.

Doesn't metadata server follow that pattern as well? It has a `Run()`, etc. It _could_ be called from zedbox, as it meets the interface requirements. It isn't, as we call it from inside zedrouter. But it still fits the design pattern?

uncleDecart replied:
It does, except for folder structure

deitch replied:
Given how long it takes for CI to run, I will leave this for now. I am happy to open a tiny PR just for adding this to docs afterwards.

uncleDecart reviewed (approved):
LGTM, good job @deitch !

deitch commented:
I cannot tell if all of the necessary eden tests ran or not. @eriknordmark or @uncleDecart ?

uncleDecart commented:
@eriknordmark , could you please approve this PR? I have a feeling tests could've been skipped because I don't have right permissions for EVE repo

deitch commented:
@uncleDecart I managed to kick them off, but they all fail. See https://github.com/lf-edge/eve/actions/runs/9300467391

uncleDecart commented:
@deitch I guess bumping version didn't work as planned see [here](https://github.com/lf-edge/eden/actions/runs/9284266545/job/25546308638#step:3:718); Could be that Eden 0.9.7 is broken

deitch commented:
Uh oh.

What can I do to help fix it?

uncleDecart commented:
@deitch starting commits from 28th May ([link](https://github.com/lf-edge/eden/commits/master/)) with updating dependencies broke something. Basically we need to go through tests and see where it got wrong. I see oras thing there as well [here](https://github.com/lf-edge/eden/commits/master/)

eriknordmark commented:
@uncleDecart one of the failing Eden tests shows this log output - any idea whether the Bad Request is something from inside Eden or Adam?

> eden eve onboard
edenProg: /home/runner/actions-runner/_work/eve/eve/eden/dist/bin/eden timewait: 0s
time="2024-05-31T07:04:04Z" level=info msg="Test is running: 2024-05-31T07:04:04Z"
Error: [stdout]
time="2024-05-31T07:02:24Z" level=info msg="Adam waiting for EVE registration (0) of (20)"
time="2024-05-31T07:02:44Z" level=info msg="Adam waiting for EVE registration (1) of (20)"
time="2024-05-31T07:03:04Z" level=info msg="Adam waiting for EVE registration (2) of (20)"
time="2024-05-31T07:03:24Z" level=info msg="Device uuid: f1e20bea-9d63-40d9-a262-f4fe0ab25a9c"
config changed, to see config run 'eden controller edge-node get-config'
time="2024-05-31T07:03:25Z" level=info msg="Received unexpected StatusCode(Bad Request): repeat request (0) of (20)"
time="2024-05-31T07:03:30Z" level=info msg="Received unexpected StatusCode(Bad Request): repeat request (1) of (20)"
time="2024-05-31T07:03:35Z" level=info msg="Received unexpected StatusCode(Bad Request): repeat request (2) of (20)"
...

giggsoff commented:
> @uncleDecart one of the failing Eden tests shows this log output - any idea whether the Bad Request is something from inside Eden or Adam?
> 
> > eden eve onboard
> > edenProg: /home/runner/actions-runner/_work/eve/eve/eden/dist/bin/eden timewait: 0s
> > time="2024-05-31T07:04:04Z" level=info msg="Test is running: 2024-05-31T07:04:04Z"
> > Error: [stdout]
> > time="2024-05-31T07:02:24Z" level=info msg="Adam waiting for EVE registration (0) of (20)"
> > time="2024-05-31T07:02:44Z" level=info msg="Adam waiting for EVE registration (1) of (20)"
> > time="2024-05-31T07:03:04Z" level=info msg="Adam waiting for EVE registration (2) of (20)"
> > time="2024-05-31T07:03:24Z" level=info msg="Device uuid: f1e20bea-9d63-40d9-a262-f4fe0ab25a9c"
> > config changed, to see config run 'eden controller edge-node get-config'
> > time="2024-05-31T07:03:25Z" level=info msg="Received unexpected StatusCode(Bad Request): repeat request (0) of (20)"
> > time="2024-05-31T07:03:30Z" level=info msg="Received unexpected StatusCode(Bad Request): repeat request (1) of (20)"
> > time="2024-05-31T07:03:35Z" level=info msg="Received unexpected StatusCode(Bad Request): repeat request (2) of (20)"
> > ...

From inside of bundled logs (adam.log file to be more precise), I can see that Adam complains about no attest information (which expected from Eden side to be populated by EVE-OS).
Also in console.log file I can see several reboots initiated by watchdog on different agent (assume there are no updates on monitored files).
I did not jump into details and potentially checked artifacts of old runs (sorry for that), but it looks like a problem on EVE-OS side.

deitch commented:
The only thing really changed in here is verifier. That isn't involved in any of that process, as far as I recall.

@giggsoff do you know how I can run just those failing tests? If I can recreate it offline, maybe I can dig into the issue.

eriknordmark commented:
> From inside of bundled logs (adam.log file to be more precise), I can see that Adam complains about no attest information (which expected from Eden side to be populated by EVE-OS to Adam right after onboarding).

@giggsoff I thought we only send the attest info when the device has a TPM. Does Adam get stuck if it doesan't receive the attest message(s)?

giggsoff commented:
> The only thing really changed in here is verifier. That isn't involved in any of that process, as far as I recall.
> 
> @giggsoff do you know how I can run just those failing tests? If I can recreate it offline, maybe I can dig into the issue.

Tests are failing during onboarding, so instead of tests you can just run `eden eve onboard` from the [doc.](https://github.com/lf-edge/eden/blob/master/docs/flow.md#launching-eve). Do not forget to build/pull image of EVE-OS with your changes and [set](https://github.com/lf-edge/eden/blob/master/docs/flow.md#setting-the-live-eve-image) it during setup. Hope @uncleDecart can help you with setup/teardown of Eden.

giggsoff commented:
> > From inside of bundled logs (adam.log file to be more precise), I can see that Adam complains about no attest information (which expected from Eden side to be populated by EVE-OS to Adam right after onboarding).
> 
> @giggsoff I thought we only send the attest info when the device has a TPM. Does Adam get stuck if it doesan't receive the attest message(s)?

We want to get ECDH Certificate to encrypt our data later (encrypted blocks in edge device config). And the certificate come from EVE-OS to Adam during attestation process regarless of TPM support. Not sure about current versions of EVE-OS.

eriknordmark commented:
> We want to get ECDH Certificate to encrypt our data later (encrypted blocks in edge device config). And the certificate come from EVE-OS to Adam during attestation process regarless of TPM support. Not sure about current versions of EVE-OS.

Makes sense to wait for the certifcate payload. So someone needs to look at the EVE logs to see if it isn't sending those for some reason (such as not getting OK responses to some earlier messages sent to Adam).

deitch commented:
> Tests are failing during onboarding, so instead of tests you can just run eden eve onboard from the [doc.](https://github.com/lf-edge/eden/blob/master/docs/flow.md#launching-eve). Do not forget to build/pull image of EVE-OS with your changes and [set](https://github.com/lf-edge/eden/blob/master/docs/flow.md#setting-the-live-eve-image) it during setup. Hope @uncleDecart can help you with setup/teardown of Eden

Thanks, I will try that. Need to build my variant of eve on my build server, then load it into local docker, then set it up for eden.

deitch commented:
I guess this is what you were seeing @giggsoff ?

```
# ./eden eve onboard
INFO[0000] Received unexpected StatusCode(Bad Request): repeat request (0) of (20)
INFO[0005] Received unexpected StatusCode(Bad Request): repeat request (1) of (20)
INFO[0010] Received unexpected StatusCode(Bad Request): repeat request (2) of (20)
INFO[0015] Received unexpected StatusCode(Bad Request): repeat request (3) of (20)
INFO[0020] Received unexpected StatusCode(Bad Request): repeat request (4) of (20)
INFO[0025] Received unexpected StatusCode(Bad Request): repeat request (5) of (20)
```

deitch commented:
Ok, this is even stranger:

```
# ./eden log
INFO[0000] Received unexpected StatusCode(Bad Request): repeat request (0) of (20)
INFO[0005] Received unexpected StatusCode(Bad Request): repeat request (1) of (20)
```

I don't know if that is the logs, or if the request to get logs is failing.

deitch commented:
Very suspicious:

```
# ./eden controller edge-node get-config
INFO[0000] Received unexpected StatusCode(Bad Request): repeat request (0) of (20)
```

deitch commented:
```
# curl -k https://139.178.82.203:3333/admin/device
unable to refresh certs from Redis: failed to retrieve onboarding certificated from ONBOARD_CERTS WRONGPASS invalid username-password pair or user is disabled.
```

Something is messed up in Adam.

deitch commented:
It turns out I had an older redis running. eden didn't recognize that it was different, so it kept it, but it had a different generated password. Oops.

deitch commented:
OK, so I managed to onboard the device, and then it got stuck.

```sh
# ./eden eve onboard
INFO[0000] Adam waiting for EVE registration (0) of (20)
INFO[0020] Adam waiting for EVE registration (1) of (20)
INFO[0040] Device uuid: c2ce8691-6621-430a-81e1-393417caa68d
config changed, to see config run 'eden controller edge-node get-config'
INFO[0041] Received unexpected StatusCode(Bad Request): repeat request (0) of (20)
INFO[0046] Received unexpected StatusCode(Bad Request): repeat request (1) of (20)
INFO[0051] Received unexpected StatusCode(Bad Request): repeat request (2) of (20)
INFO[0056] Received unexpected StatusCode(Bad Request): repeat request (3) of (20)
INFO[0061] Received unexpected StatusCode(Bad Request): repeat request (4) of (20)
INFO[0066] Received unexpected StatusCode(Bad Request): repeat request (5) of (20)
INFO[0071] Received unexpected StatusCode(Bad Request): repeat request (6) of (20)
INFO[0076] Received unexpected StatusCode(Bad Request): repeat request (7) of (20)
INFO[0081] Received unexpected StatusCode(Bad Request): repeat request (8) of (20)
INFO[0086] Received unexpected StatusCode(Bad Request): repeat request (9) of (20)
INFO[0091] Received unexpected StatusCode(Bad Request): repeat request (10) of (20)
INFO[0096] Received unexpected StatusCode(Bad Request): repeat request (11) of (20)
INFO[0101] Received unexpected StatusCode(Bad Request): repeat request (12) of (20)
INFO[0106] Received unexpected StatusCode(Bad Request): repeat request (13) of (20)
INFO[0111] Received unexpected StatusCode(Bad Request): repeat request (14) of (20)
INFO[0116] Received unexpected StatusCode(Bad Request): repeat request (15) of (20)
INFO[0121] Received unexpected StatusCode(Bad Request): repeat request (16) of (20)
INFO[0126] Received unexpected StatusCode(Bad Request): repeat request (17) of (20)
```

Status is:

```sh
# ./eden status
✔ Adam status: container with name eden_adam is running
	Adam is expected at https://139.178.82.203:3333
	For local Adam you can run 'docker logs eden_adam' to see logs
✔ Registry status: container with name eden_registry is running
	Registry is expected at https://139.178.82.203:5050
	For local registry you can run 'docker logs eden_registry' to see logs
✔ Redis status: container with name eden_redis is running
	Redis is expected at 139.178.82.203:6379
	For local Redis you can run 'docker logs eden_redis' to see logs
✔ EServer process status: container with name eden_eserver is running
	EServer is expected at http://139.178.82.203:8888 from EVE
	For local EServer you can run 'docker logs eden_eserver' to see logs

--- context: default ---
EVE state: registered

? EVE REMOTE IPs: waiting for info...
? EVE memory: waiting for info...
✔ EVE on Qemu status: running with pid 146648
	Logs for local EVE at: /root/go/src/github.com/lf-edge/eden/dist/default-eve.log
✔ EVE Request IP: 139.178.82.203
------
```

I can run `get-options`, etc. 

I am not sure where the `INFO[0041] Received unexpected StatusCode(Bad Request): repeat request (0) of (20)` is coming from. Somewhere inside Adam, I suspect?

deitch commented:
Run again:

```sh
# ./eden eve onboard
INFO[0000] onboarded
INFO[0000] device UUID: c2ce8691-6621-430a-81e1-393417caa68d
```

It is fine with it. So where are those errors coming from?

giggsoff commented:
> OK, so I managed to onboard the device, and then it got stuck.
> 
> ```shell
> # ./eden eve onboard
> INFO[0000] Adam waiting for EVE registration (0) of (20)
> INFO[0020] Adam waiting for EVE registration (1) of (20)
> INFO[0040] Device uuid: c2ce8691-6621-430a-81e1-393417caa68d
> config changed, to see config run 'eden controller edge-node get-config'
> INFO[0041] Received unexpected StatusCode(Bad Request): repeat request (0) of (20)
> INFO[0046] Received unexpected StatusCode(Bad Request): repeat request (1) of (20)
> INFO[0051] Received unexpected StatusCode(Bad Request): repeat request (2) of (20)
> INFO[0056] Received unexpected StatusCode(Bad Request): repeat request (3) of (20)
> INFO[0061] Received unexpected StatusCode(Bad Request): repeat request (4) of (20)
> INFO[0066] Received unexpected StatusCode(Bad Request): repeat request (5) of (20)
> INFO[0071] Received unexpected StatusCode(Bad Request): repeat request (6) of (20)
> INFO[0076] Received unexpected StatusCode(Bad Request): repeat request (7) of (20)
> INFO[0081] Received unexpected StatusCode(Bad Request): repeat request (8) of (20)
> INFO[0086] Received unexpected StatusCode(Bad Request): repeat request (9) of (20)
> INFO[0091] Received unexpected StatusCode(Bad Request): repeat request (10) of (20)
> INFO[0096] Received unexpected StatusCode(Bad Request): repeat request (11) of (20)
> INFO[0101] Received unexpected StatusCode(Bad Request): repeat request (12) of (20)
> INFO[0106] Received unexpected StatusCode(Bad Request): repeat request (13) of (20)
> INFO[0111] Received unexpected StatusCode(Bad Request): repeat request (14) of (20)
> INFO[0116] Received unexpected StatusCode(Bad Request): repeat request (15) of (20)
> INFO[0121] Received unexpected StatusCode(Bad Request): repeat request (16) of (20)
> INFO[0126] Received unexpected StatusCode(Bad Request): repeat request (17) of (20)
> ```
> 
> Status is:
> 
> ```shell
> # ./eden status
> ✔ Adam status: container with name eden_adam is running
> 	Adam is expected at https://139.178.82.203:3333
> 	For local Adam you can run 'docker logs eden_adam' to see logs
> ✔ Registry status: container with name eden_registry is running
> 	Registry is expected at https://139.178.82.203:5050
> 	For local registry you can run 'docker logs eden_registry' to see logs
> ✔ Redis status: container with name eden_redis is running
> 	Redis is expected at 139.178.82.203:6379
> 	For local Redis you can run 'docker logs eden_redis' to see logs
> ✔ EServer process status: container with name eden_eserver is running
> 	EServer is expected at http://139.178.82.203:8888 from EVE
> 	For local EServer you can run 'docker logs eden_eserver' to see logs
> 
> --- context: default ---
> EVE state: registered
> 
> ? EVE REMOTE IPs: waiting for info...
> ? EVE memory: waiting for info...
> ✔ EVE on Qemu status: running with pid 146648
> 	Logs for local EVE at: /root/go/src/github.com/lf-edge/eden/dist/default-eve.log
> ✔ EVE Request IP: 139.178.82.203
> ------
> ```
> 
> I can run `get-options`, etc.
> 
> I am not sure where the `INFO[0041] Received unexpected StatusCode(Bad Request): repeat request (0) of (20)` is coming from. Somewhere inside Adam, I suspect?

It is Eden trying to get attest data from Adam. Adam, in turns, get it from EVE-OS attestation process. You can try to check additional logs of Adam in `docker logs eden_adam`. Also you can take a look onto `dist/default-eve.log` file to get some information from EVE-OS console (try to grep by word `Rebooting` firstly).

deitch commented:
The `default-eve.log` file is a few thousand lines of EVE booting and rebooting about 6 times. Not a lot interesting there, possibly except for the following, which I don't think matters:

```
INFO: device: unspecified attest: None vault: UNKNOWN pcr: UNKNOWN^M
```

The logs for Adam have a little more. This repeats a lot:

```
2024/06/05 17:25:27 139.178.82.203:51816 (CN=EVE,O=The Linux Foundation) requested /api/v2/edgedevice/uuid
2024/06/05 17:25:27 error checking DeviceCheckCertHash: cert hash not found
2024/06/05 17:25:27 139.178.82.203:51830 (CN=EVE,O=The Linux Foundation) requested /api/v2/edgedevice/uuid
2024/06/05 17:25:27 error checking DeviceCheckCertHash: cert hash not found
2024/06/05 17:25:27 139.178.82.203:51840 (CN=EVE,O=The Linux Foundation) requested /api/v2/edgedevice/uuid
2024/06/05 17:25:27 error checking DeviceCheckCertHash: cert hash not found
2024/06/05 17:25:27 139.178.82.203:51848 (CN=EVE,O=The Linux Foundation) requested /api/v2/edgedevice/uuid
2024/06/05 17:25:27 error checking DeviceCheckCertHash: cert hash not found
2024/06/05 17:25:29 139.178.82.203:51860 (CN=39f2887c-b205-42ba-8535-f64184e609ff,O=lf-edge,C=RU) requested /api/v2/edgedevice/register
2024/06/05 17:25:29 failed to onboard with serial: unknown onboarding certificate
2024/06/05 17:25:29 will retry with soft serial
2024/06/05 17:25:29 Failed in registerProcess: failed authentication unknown onboarding certificate
2024/06/05 17:25:29 139.178.82.203:51862 (CN=39f2887c-b205-42ba-8535-f64184e609ff,O=lf-edge,C=RU) requested /api/v2/edgedevice/register
2024/06/05 17:25:29 failed to onboard with serial: unknown onboarding certificate
2024/06/05 17:25:29 will retry with soft serial
2024/06/05 17:25:29 Failed in registerProcess: failed authentication unknown onboarding certificate
2024/06/05 17:25:29 139.178.82.203:51872 (CN=EVE,O=The Linux Foundation) requested /api/v2/edgedevice/uuid
2024/06/05 17:25:29 error checking DeviceCheckCertHash: cert hash not found
```

eventually it ends in:

```
2024/06/05 17:26:25 139.178.82.203:48076 (CN=39f2887c-b205-42ba-8535-f64184e609ff,O=lf-edge,C=RU) requested /api/v2/edgedevice/register
2024/06/05 17:26:26 139.178.82.203:48082 (CN=EVE,O=The Linux Foundation) requested /api/v2/edgedevice/uuid
2024/06/05 17:26:43 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:26:48 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:26:53 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:26:58 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:27:03 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:27:08 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:27:13 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:27:18 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:27:23 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:27:28 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:27:33 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:27:38 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:27:43 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:27:48 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:27:53 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:27:58 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:28:01 http: TLS handshake error from 87.79.22.198:60220: remote error: tls: illegal parameter
2024/06/05 17:28:03 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:28:08 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:28:13 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:28:18 deviceCertsGet: failed to get attest for c2ce8691-6621-430a-81e1-393417caa68d: redis: nil
2024/06/05 17:40:40 http: TLS handshake error from 45.123.27.69:45522: remote error: tls: illegal parameter
```

giggsoff commented:
> The `default-eve.log` file is a few thousand lines of EVE booting and rebooting about 6 times. Not a lot interesting there, possibly except for the following, which I don't think matters:
> 
> ```
> INFO: device: unspecified attest: None vault: UNKNOWN pcr: UNKNOWN^M
> ```

But did you fire a reboot or something like this? In GH Actions it was the main problem, rebooting because of watchdog (something like `Rebooting EVE. Reason: Watchdog report`, can you see such lines?), which fires imidiately after onboard before attestation (for sure I did not check timings, it is my assumption).

deitch commented:
Sure.

```
Rebooting EVE. Reason: Watchdog report for IMGA EVE version 0.0.0-restructure-verifier-a7991fc2-kvm-amd64 at 2024-06-05T17:27:49+00:00: 3 /run/zedbox.pid
Rebooting EVE. Reason: Watchdog report for IMGA EVE version 0.0.0-restructure-verifier-a7991fc2-kvm-amd64 at 2024-06-05T17:31:32+00:00: 2 /run/zedrouter.touch
Rebooting EVE. Reason: Watchdog report for IMGA EVE version 0.0.0-restructure-verifier-a7991fc2-kvm-amd64 at 2024-06-05T17:46:56+00:00: 2 /run/zedkube.touch
Rebooting EVE. Reason: Watchdog report for IMGA EVE version 0.0.0-restructure-verifier-a7991fc2-kvm-amd64 at 2024-06-05T17:57:48+00:00: 2 /run/zedkube.touch
Rebooting EVE. Reason: Watchdog report for IMGA EVE version 0.0.0-restructure-verifier-a7991fc2-kvm-amd64 at 2024-06-05T18:06:36+00:00: 2 /run/usbmanager.touch
```

deitch commented:
It sounds like you are saying EVE keeps rebooting due to watchdog issues. That has to be figured out and fixed. OK.

What does that have to do with those `INFO[0041] Received unexpected StatusCode(Bad Request)` messages from adam?

giggsoff commented:
> What does that have to do with those `INFO[0041] Received unexpected StatusCode(Bad Request)` messages from adam? And isn't the device actually onboarded?

It comes from the step in `eden eve onboard` logic: [here](https://github.com/lf-edge/eden/blob/a657ac011c03527ef11d47036dbc793fdadb77a2/pkg/controller/functions.go#L169). Device is onboarded on that step, but Eden also want to get the ECDH certificate.

giggsoff commented:
> So is the problem that line, ie it is failing to get the cert? Or is it the watchdog reboots? Or both? Or are they related?

The problem shown in tests is about failing to get the cert. But why Eden cannot get it? Because EVE-OS did not put it inside Adam (at least during the test, that is several minutes). Why EVE-OS did not put it? It is a good question I have no answer for. I just can see the unexpected reboots in console and assume that it is related.

Potentially it may be the problem in Eden/Adam, e.g. for some reason Adam forgot how to receive, store and respond cert properly. If you have a time, you can just try to run the same versions of Eden/Adam with previous version of EVE-OS to double check.

eriknordmark commented:
> The problem shown in tests is about failing to get the cert. But why Eden cannot get it? Because EVE-OS did not put it inside Adam (at least during the test, that is several minutes). Why EVE-OS did not put it? It is a good question I have no answer for. I can see the unexpected reboots in console and assume that they are related.

@giggsoff I don't understand how the log.Fatal is supposed to work. The certificate might arrive a few milliseconds or a few seconds after the device has been onboarded (completed the /register API call) with Adam. So shouldn't Adam wait for the Cert to appear and not Fatal?

giggsoff commented:
> > The problem shown in tests is about failing to get the cert. But why Eden cannot get it? Because EVE-OS did not put it inside Adam (at least during the test, that is several minutes). Why EVE-OS did not put it? It is a good question I have no answer for. I can see the unexpected reboots in console and assume that they are related.
> 
> @giggsoff I don't understand how the log.Fatal is supposed to work. The certificate might arrive a few milliseconds or a few seconds after the device has been onboarded (completed the /register API call) with Adam. So shouldn't Adam wait for the Cert to appear and not Fatal?

Adam in that case wait for requests of EVE-OS and do not do any log.Fatal. Adam stores the cert when it comes (does not matter, in seconds or days after onboarding).
But Eden cannot wait forever as it try to say in some limited time that EVE-OS is potentially broken and we should stop the test. Eden has [retry](https://github.com/lf-edge/eden/blob/a657ac011c03527ef11d47036dbc793fdadb77a2/pkg/utils/networking.go#L170) logic for requests to Adam, we can see in logs that we tried several times (with delay), and exits only if all of them was errored.

deitch commented:
I now have 100 such "Rebooting" lines since I launched it yesterday. Clearly, EVE is unhappy and keeps rebooting. That might to might not be related to the issue. But it is one thing to track down.

>> What does that have to do with those INFO[0041] Received unexpected StatusCode(Bad Request) messages from adam? And isn't the device actually onboarded?
>
> It comes from the step in eden eve onboard logic: [here](https://github.com/lf-edge/eden/blob/a657ac011c03527ef11d47036dbc793fdadb77a2/pkg/controller/functions.go#L169). Device is onboarded on that step, but Eden also want to get the ECDH certificate.

I can see that, but the error message is one of those "INFO[0041] Received unexpected StatusCode(Bad Request)". That doesn't sound like, "I reached Adam, I made a request, it didn't have what I wanted." Those errors would have had the messages in [GetECDHCert()](https://github.com/lf-edge/eden/blob/master/pkg/controller/adam/adam.go#L203), like "cannot get attestation certificates from cloud" or "cannot unmarshal attest" or "no DEVICE_ECDH_EXCHANGE certificate". Instead, we are seeing those messages from `networking.go`. 

Or are you saying that those are the `log.Infof()` messages on the CLI that we are seeing, but the errors themselves are swallowed somewhere? If we had those, this would be much easier to debug.

deitch commented:
I stepped through `eve onboard` with a debugger. The URL it is trying to get to is `/admin/device/5644d68f-68d3-487f-bce4-82b60a61c2d8/certs`, the response is a 400.

As far as I can tell, there is no way to remove a device via `eden controller`? I don't want to stop and clear everything, just remove the device from the controller and create a new device?

deitch commented:
Ah, docker logs for Adam:

```
2024/06/06 07:26:45 deviceCertsGet: failed to get attest for 5644d68f-68d3-487f-bce4-82b60a61c2d8: redis: nil
2024/06/06 07:31:09 deviceCertsGet: failed to get attest for 5644d68f-68d3-487f-bce4-82b60a61c2d8: redis: nil
2024/06/06 07:31:13 deviceCertsGet: failed to get attest for 5644d68f-68d3-487f-bce4-82b60a61c2d8: redis: nil
```

That probably shouldn't return a 400, it should return something else. And eden should handle it better.

deitch commented:
Did some more digging. The failure in Adam is [here](https://github.com/lf-edge/adam/blob/master/pkg/driver/redis/device_manager_redis.go#L657-L668), which returns an error. The root issue is, as @giggsoff said, the fact that the attestation cert never is uploaded, and thus does not end up in redis.

When it does not succeed in getting it, it then returns that error to [here](https://github.com/lf-edge/adam/blob/master/pkg/server/adminHandler.go#L540). The logic _should_ return a 404 if the error is a `NotFoundError`, but instead it is returning a 400:

```go
	deviceAttest, err := h.manager.GetCerts(uid)
	_, isNotFound := err.(*common.NotFoundError)
	switch {
	case err != nil && isNotFound:
		log.Printf("deviceCertsGet: %v", err)
		http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
	case err != nil:
		log.Printf("deviceCertsGet: %v", err)
		http.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest)
	case deviceAttest == nil:
		http.Error(w, "found device information, but certs was empty", http.StatusInternalServerError)
	default:
		w.WriteHeader(http.StatusOK)
		w.Write(deviceAttest)
	}
```

This is because the redis driver (and maybe others) does not distinguish between not found and error [here](https://github.com/lf-edge/adam/blob/master/pkg/driver/redis/device_manager_redis.go#L657-L668):

```go
func (d *DeviceManager) GetCerts(u uuid.UUID) ([]byte, error) {
	// hold our config
	var b []byte
	data, err := d.client.HGet(deviceAttestCertsHash, u.String()).Result()
	if err != nil {
		return nil, fmt.Errorf("failed to get attest for %s: %v", u.String(), err)
	} else {
		b = []byte(data)
	}

	return b, nil
}
```

From the caller's perspective, it just looks like an error. Fixing this would not solve our real issue with EVE, but it would fix eden misunderstanding the result.

deitch commented:
First fix (for Adam) [here](https://github.com/lf-edge/adam/pull/126)

deitch commented:
Slowly. I managed to get onto the device briefly, looked at `/persist/newlog/dev.*`:

```json
{"devID":"53cbe8d3-f298-449f-b9ab-e2bc96bd0a2b","image":"IMGA","eveVersion":"0.0.0-restructure-verifier-a7991fc2-kvm-amd64"}
{"severity":"err","source":"watchdog","iid":"13956","content":"pinging process 1491 (/run/zedbox.pid) gave errno = 3 = 'No such process'","msgid":2065,"timestamp":{"seconds":1717665386,"nanos":256940686}}
{"source":"watchdog","content":"watchdog: pinging process 1491 (/run/zedbox.pid) gave errno = 3 = 'No such process'","msgid":2066,"timestamp":{"seconds":1717665386,"nanos":257000000}}
{"severity":"err","source":"watchdog","iid":"13956","content":"pinging process 1491 (/run/zedbox.pid) gave errno = 3 = 'No such process'","msgid":2067,"timestamp":{"seconds":1717665387,"nanos":257399812}}
{"source":"watchdog","content":"watchdog: pinging process 1491 (/run/zedbox.pid) gave errno = 3 = 'No such process'","msgid":2068,"timestamp":{"seconds":1717665387,"nanos":257445529}}
{"severity":"err","source":"watchdog","iid":"13956","content":"pinging process 1491 (/run/zedbox.pid) gave errno = 3 = 'No such process'","msgid":2069,"timestamp":{"seconds":1717665388,"nanos":257849155}}
{"source":"watchdog","content":"watchdog: pinging process 1491 (/run/zedbox.pid) gave errno = 3 = 'No such process'","msgid":2070,"timestamp":{"seconds":1717665388,"nanos":257916234}}
{"severity":"err","source":"watchdog","iid":"13956","content":"pinging process 1491 (/run/zedbox.pid) gave errno = 3 = 'No such process'","msgid":2071,"timestamp":{"seconds":1717665389,"nanos":258374428}}
```

europaul reviewed (approved):
very nice refactoring!

deitch commented on pkg/pillar/cmd/README.md:
```diff
@@ -0,0 +1,273 @@
+
+In addition, it should contain the following subdirectories:
+
+* [`lib/`](./lib/) contains core logic, has no dependency on being called command-line or pubsub, unless pubsub is part of the core business logic. It is important to keep these functions as simple and straightforward as possible. Unit tests of business logic should be here as well.
```
A very good point. I will fix that. First, I want to see if we can get this to pass tests.

deitch replied:
Fixed

eriknordmark commented:
@deitch doing make live run 
and then onboarding the device to the controller using /config/device.cert.pem, I see the following in the logs:
panic: interface conversion: interface {} is *verifier.Verifier, not *pubsub.VerifierContext

goroutine 197 [running]:
github.com/lf-edge/eve/pkg/pillar/cmd/verifier/pubsub.(*VerifierContext).handleG
lobalConfigImpl(0xc000f19370?, {0x29acb60?, 0x0?}, {0xc0024327f0?, 0xc0024327f0?
}, {0x6?, 0x289d960?})
	/pillar/cmd/verifier/pubsub/handlers.go:236 +0x1c5
github.com/lf-edge/eve/pkg/pillar/cmd/verifier/pubsub.(*VerifierContext).handleG
lobalConfigCreate(...)
	/pillar/cmd/verifier/pubsub/handlers.go:225
github.com/lf-edge/eve/pkg/pillar/pubsub.handleModify({0x2a46b00?, 0xc00111f200}
, {0xc0024327f0, 0x6}, {0xc0011e0000, 0x1f56, 0x1f57})
	/pillar/pubsub/subscribe.go:269 +0x735
github.com/lf-edge/eve/pkg/pillar/pubsub.(*SubscriptionImpl).populate(0xc00111f2
00)
	/pillar/pubsub/subscribe.go:97 +0x227
github.com/lf-edge/eve/pkg/pillar/pubsub.(*SubscriptionImpl).Activate(0xc00111f2
00)
	/pillar/pubsub/subscribe.go:54 +0x25
github.com/lf-edge/eve/pkg/pillar/cmd/verifier/pubsub.(*VerifierContext).Run(0xc
000e46270, {0x45cf5d0?, 0x0?, 0x0?}, {0x0?, 0x0?})

deitch commented:
Yeah, that was it. I now have it booting, `eden eve onboard` worked, `eden eve console` works, `eden eve ssh` works.

OK, all of that. It shouldn't have been that hard and taken so many people, but I thank you all.

@eriknordmark we want to kick off eden tests again on this?

eriknordmark reviewed (changes requested):
Note that there are 10 yetus issues reported, which probably means that there are more than 10 (since it reports at most 10). So please fix them (or if some are not fixable, add the comments to silence those).

It was this one which might point at an issue with the type assertions:
[yetus: pkg/pillar/cmd/verifier/pubsub/handlers.go#L289](https://github.com/lf-edge/eve/pull/3954/files#annotation_22495285694)
golangcilint: type assertion must be checked (forcetypeassert)

deitch commented:
@eriknordmark where do you see yetus annotations? I went to the [files](https://github.com/lf-edge/eve/pull/3954/files), I see exactly one yetus annotation, that is on quoted text inside a doc, which we need to ignore. How did you find 10?

giggsoff commented:
> @eriknordmark where do you see yetus annotations? I went to the [files](https://github.com/lf-edge/eve/pull/3954/files), I see exactly one yetus annotation, that is on quoted text inside a doc, which we need to ignore. How did you find 10?

It looks like they are hidden by `Load diff` button. Not so convenient.

![image](https://github.com/lf-edge/eve/assets/5590137/336995c3-a8a2-4cce-961e-4968e0149260)

deitch commented:
Yeah. It really would be nice if, in addition to the annotations, it did a full report in a single screen.

giggsoff commented:
> Yeah. It really would be nice if, in addition to the annotations, it did a full report in a single screen.

Well, it is also [available.](https://github.com/lf-edge/eve/actions/runs/9404998433). It is summary for yetus workflow.

deitch commented:
One yetus complaint left about a quoted section.

eriknordmark reviewed (approved):
Kick off tests (still need to review the documentation)

deitch commented:
Lots of eden failures. Going to rebase on latest `master` and push out again. That will need a new approval eventually.

eriknordmark commented on pkg/pillar/cmd/README.md:
```diff
@@ -0,0 +1,273 @@
+
+In addition, it should contain the following subdirectories:
+
+* `lib/` contains core logic, has no dependency on being called command-line or pubsub, unless pubsub is part of the core business logic. It is important to keep these functions as simple and straightforward as possible. Unit tests of business logic should be here as well.
+* `pubsub/` contains functions to handle pubsub messages, including all handlers. There should be no business logic here, only pubsub message handling. For business logic, handlers should call business logic functions in `lib/`.
```
For things like the metadata cmd, where all the business logic is to pass between pubsub and a http server, what do we recommend?

 Put all of it in lib/, all in pubsub/, or rework so that the business logic calls to pubsub Get and Publish are replaced by calling some internal functions which can be stubbed out for testing?

FWIW same question will apply to pubsub-heaving microservices in zedmanager.

deitch replied:
This is a fair point. I don't have an obvious answer. My immediate one is, this is iterative. My intent is, once this works, we extend it to the bulk of other services which do not use pubsub as core business logic, but rather as a messaging system. Once those are done, we can attack those.

Or we can ask @uncleDecart to come up with a good solution, since he is doing the independent metadata server. 😁

eriknordmark commented on pkg/pillar/cmd/verifier/cmd/oneshot.go:
```diff
@@ -0,0 +1,64 @@
+package main
```
Please add Copyright and License lines to all new files.

deitch replied:
Oops. Yeah, adding now.

deitch replied:
Done. Thank you for reminding me.

eriknordmark commented:
The diffs show three files being deleted when in fact they have been moved to a subdir. Can you see if you can get them to show as diffs so they can be reviewed?

deitch commented:
Any idea how?

giggsoff commented:
> Any idea how?

Did not checked, but looks like you can try to split the changes into two commits: first one is just moving files and the second one is change in them. I hope it will allow reviewers to filter by commits (GH allow to do it) and review diff in the one that has a change.

deitch commented:
I tried to split it into multiple commits. I am not sure I didn't break something along the way. If it did, I still have a local copy of the original 2-commit change, rather than this one. I don't know if it will help @eriknordmark but we can hope.

Be aware that the first 3 commits will not work standalone:

* moved some files around and added some new ones
* redid the moved files
* added package changes like go.mod, vendor, etc.

The other 2 commits are fine on their own:

* golangci.yml exclusions (latest versions of golangci-lint include mnd which will annoy us no end)
* document how to structure agents after the changes

back to you @eriknordmark

eriknordmark commented on pkg/pillar/cmd/verifier/lib/logger.go:
```diff
@@ -0,0 +1,18 @@
+
+package verifier
+
+// Logger is an interface for logging, combines the best of logrus.Logger and base.LogObject
```
If this is useful should we later promote it (e.g., to base.Logger)? No need to touch this PR though.

deitch replied:
Yeah, that would be a good thing to do. Once this one is in, I will open a separate PR just for this.

I have used this pattern in the past. I never really understood why `log.Logger` (or similar) is not an interface. Just work with what we have, I guess.

eriknordmark commented on pkg/pillar/cmd/verifier/pubsub/context.go:
```diff
@@ -0,0 +1,191 @@
+package pubsub
```
Hmm - copyright and license missing in this file.

deitch replied:
Done

eriknordmark commented on pkg/pillar/cmd/verifier/pubsub/const.go:
```diff
@@ -0,0 +1,18 @@
+package pubsub
```
Ditto

deitch replied:
Darn, I thought I had gotten these. I will do another quick check.

deitch replied:
Done.

eriknordmark reviewed (approved):
Thanks for splitting the commits.
LGTM but some missing copyright/license files.

deitch commented:
> Thanks for splitting the commits.

Did that work? Did it make it easier to read?

> LGTM but some missing copyright/license files.

Fixed them. I do not think it needs another approval. Now to see what eden does.

eriknordmark reviewed (approved):
Run eden again

christoph-zededa commented on pkg/pillar/cmd/verifier/README.md:
```diff
@@ -0,0 +1,26 @@
+)
+
+func main() {
+    verifier.Run(pubSubImpl, logrusLogger, baseLogObject, []string{argument1, argument2})
```
What about having this instead in a golang example? (https://go.dev/blog/examples )
Then we now if we break this code.

deitch replied:
You know, I rather like that idea. But forgive me, I am going to defer on that to a separate PR. It has been so long and hard to get this in, especially with all of the eden issues, that I want to get this in as we can. Once it is in, a separate PR with just some doc+example changes will be nice and light.

christoph-zededa replied:
Fine :-)

milan-zededa commented on pkg/pillar/cmd/README.md:
```diff
@@ -0,0 +1,273 @@
+
+In addition, it should contain the following subdirectories:
+
+* `lib/` contains core logic, has no dependency on being called command-line or pubsub, unless pubsub is part of the core business logic. It is important to keep these functions as simple and straightforward as possible. Unit tests of business logic should be here as well.
```
What if microservice is bigger and consists of multiple components (each implemented as a separate package)?
For example, [DPC Reconciler](https://github.com/lf-edge/eve/tree/master/pkg/pillar/dpcreconciler) is a component of the NIM microservice. Does it mean that it should be moved to `pillar/cmd/nim/lib/dpcreconciler`? I'm afraid it would lead to a very deeply nested directory structure.

deitch replied:
Why should it move? As it stands now, `dpcreconciler` is a library that is pulled in by NIM. Why should it have to move? The focus here is entirely on the entrypoints for agents and their structure.

Using your specific example, I would see nothing changing for `dpcreconciler`, while NIM would (eventually) move to a `lib/`, `pubsub/`, `cmd/` layout.

milan-zededa commented on pkg/pillar/cmd/README.md:
```diff
@@ -0,0 +1,273 @@
+The package name in the agent's directory is the same as the directory name.
+It should contain at least the following files:
+
+* `README.md` describing the agent and its purpose.
```
Are we eventually going to get rid of https://github.com/lf-edge/eve/tree/master/pkg/pillar/docs? Currently the per-agent documentation is stored there.

deitch replied:
I didn't think about that. I could go either way: store docs about each service inside the service directory, or keep them in the directory https://github.com/lf-edge/eve/tree/master/pkg/pillar/docs

I have no preference, so it isn't really fundamental to this PR restructure.

What are your thoughts?

milan-zededa replied:
I actually do not have preference either :)

OhmSpectator commented:
@eriknordmark, I will review the pull request tomorrow! Thank you for bringing it to my attention!

milan-zededa commented on pkg/pillar/cmd/README.md:
```diff
@@ -0,0 +1,273 @@
+The `context.Run()` function contains all of the initialization and
+pubsub interaction.
+
+## CLI functions
```
We should probably create some dev container for running these CLIs because most microservices have some dependencies on installed software (dhcpcd, dnsmasq, etc.) and could make some config changes and break something in the host.

deitch replied:
Those are exactly the dependencies I want to see. But not all of them have it. So if you are working with downloader and need only an http server or registry for testing, you should be able to test with just those, either listed or (even better, as you recommended) a devcontainer. If you are working with NIM, you might need dhcpcd and dnsmasq, etc.

As we get to those with greater complexities, we will start to draw those out. I really like the devcontainer approach.

milan-zededa reviewed (approved):
I like this approach a lot, though it will be interesting to see how it will go with more complex microservices.

deitch commented:
> Much easier. Did you resolve the doc directory question?

If you mean the links in the doc, then, yes. If you mean, should the docs be in each agent's directory or the centralized docs, no, mostly punting on it for now. I don't think it matters for getting this one through.

deitch commented:
> it will be interesting to see how it will go with more complex microservices

It will. There are many motivations for this restructure. One motivation is to force those complexities out into the open. Just using verifier - a pretty simple service - and seeing its dependencies, as well as trying to compile using straight go on both linux and macOS, I found all sorts of interesting unexpected dependencies that affected file sizes and compiling and running. Those all were cleared up in earlier PRs.

deitch commented:
I am at a loss as to what to do next to move this along. Are the eden test failures fundamental or just issues with eden? Have there been fixes on which I should rebase this?

deitch commented:
Now I have to apply it to some other services.
